ssl/ssl_lib.c

```diff
@@ -4343,6 +4343,80 @@ const CTLOG_STORE *SSL_CTX_get0_ctlog_store(const SSL_CTX *ctx)
     return ctx->ctlog_store;
 }
 
+void
+SSL_CTX_set_early_cb(SSL_CTX *c, ssl_early_cb_fn cb, void *arg)
+{
+    c->early_cb = cb;
+    c->early_cb_arg = arg;
+}
+
+int
+SSL_EARLY_isv2(SSL *s, SSL_EARLY *ectx)
+{
+    return ectx->isv2;
+}
+
+unsigned int
+SSL_EARLY_get0_version(SSL *s, SSL_EARLY *ectx)
+{
+    return ectx->legacy_version;
+}
+
+size_t
+SSL_EARLY_get0_random(SSL *s, SSL_EARLY *ectx, const unsigned char **out)
+{
+    if (out != NULL)
+        *out = ectx->random;
+    return ectx->isv2 ? 0 : SSL3_RANDOM_SIZE;
```

Comment:
Why is it necessary to distinguish whether it is v2 here or not? If it is v2 then we use the variable length challenge as the random. If it is longer than SSL3_RANDOM_SIZE then we just use the first SSL3_RANDOM_SIZE bytes. If it is less than SSL3_RANDOM_SIZE then we pad to the left with 0s until it is SSL3_RANDOM_SIZE in length. Either way the resulting length is SSL3_RANDOM_SIZE in all cases.